src/main/java/org/reactivestreams/servlet/NonBlockingMutexExecutor.java

```diff
@@ -32,75 +40,36 @@
  * operations submitted recursively will run sequentially.
  */
 class NonBlockingMutexExecutor implements Executor {
-
-  /**
-   * We use an array for a queue, since the JDK doesn't provide any built in
-   * persistent collections, and these queues are generally going to be short enough
-   * that array copies won't be expensive (and will probably be cheaper than a
-   * persistent collection anyway).
-   */
-  private final AtomicReference<Runnable[]> state = new AtomicReference<>(null);
-
-  private static final Runnable[] empty = new Runnable[0];
+  private final AtomicReference<RunNode> last = new AtomicReference<>();
 
   @Override
   public void execute(Runnable command) {
-    Runnable[] prevState;
-    Runnable[] newState;
-
-    do {
-      prevState = state.get();
-      if (prevState == null) {
-        newState = empty;
-      } else {
-        newState = new Runnable[prevState.length + 1];
-        System.arraycopy(prevState, 0, newState, 0, prevState.length);
-        newState[prevState.length] = command;
-      }
-    } while (!state.compareAndSet(prevState, newState));
-
-    if (prevState == null) {
-      // We changed from null to a list of ops, that's mean it's our responsibility to run it
-      executeAll(command);
-    }
-
+    final RunNode newNode = new RunNode(Objects.requireNonNull(command, "Runnable must not be null"));
+    final RunNode prevLast = last.getAndSet(newNode);
+    // We changed from null to a list of ops, that's mean it's our responsibility to run it
+    if (prevLast == null) {
+      executeAll(newNode);
+    } else
+      prevLast.lazySet(newNode);
```

Comment:
Let me just confirm that I've understood how this is going to work.

Before this statement gets executed, the next task of the previous node is going to be null, and `executeAll` could read `null` when it does `next = prev.get()`. This is ok though, because the exit condition of that loop is not just whether the next task is null, it's whether the next task is null *and* the last task is the task just executed (tested using the `compareAndSet`), which it won't be because this just set it to be the new node. And so `executeAll` is going to spin on that check until this statement executes and the next task is set to the new node, at which point it will get it, and progress to make it to the current task to execute, execute it, and so on.

Reply:
@jroper You're very much correct. The lazySet will make sure that the value will eventually be published safely. (which is all we can guarantee anyway since the currently executing thread might stall for N cycles until the method is called anyway)